torchdata/datapipes/iter/transform/bucketbatcher.py

```diff
@@ -92,7 +103,104 @@ def __new__(
         if sort_key is not None:
             # In-batch shuffle each bucket seems not that useful, it seems misleading since .batch is called prior.
             if in_batch_shuffle:
-                datapipe = datapipe.batch(batch_size=bucket_num, drop_last=False).map(fn=_in_batch_shuffle_fn).unbatch()
+                datapipe = datapipe.batch(batch_size=bucket_num, drop_last=False).in_batch_shuffle().unbatch()
             else:
                 datapipe = datapipe.shuffle(buffer_size=bucket_size)
         return datapipe
+
+
+def _default_len_fn(token):
+    return len(token)
+
+
+def _token_len_fn(token, len_fn):
+    return len_fn(token), token
+
+
+def _token_min_filter_fn(data, *, min_len):
+    length, token = data
+    if length < min_len:
+        return False
+    return True
+
+
+def _token_max_filter_fn(data, *, max_len):
+    length, token = data
+    if length > max_len:
+        return False
+    return True
+
+
+@functional_datapipe("max_token_bucketize")
+class MaxTokenBucketizerIterDataPipe(IterDataPipe[DataChunk[T_co]]):
+    r"""
+    Iterable DataPipe to create mini-batches of data from bucket, and the
+    total length of each batch will be limited by ``max_token_count``.
+    (functional name: ``max_token_bucketize``).
+
+    Args:
+        datapipe: Iterable DataPipe being batched
+        max_token_count: Maximum length of total length of data in each batch
+        len_fn: Function to be applied to each element to get lengths. ``len(data)`` is used by default.
+        min_len: Optional minimum length to be included into each batch
+        max_len: Optional maximum length to be included into each batch.
+        buffer_size: This restricts how many tokens are taken from prior DataPipe to bucketize
+    """
+    datapipe: IterDataPipe[T_co]
+    max_token_count: int
+    len_fn: Callable
+    min_len: int
+    max_len: Optional[int]
+    buffer_size: int
+
+    def __init__(
+        self,
+        datapipe: IterDataPipe[T_co],
+        max_token_count: int,
+        len_fn: Callable = _default_len_fn,
+        min_len: int = 0,
+        max_len: Optional[int] = None,
+        buffer_size: int = 1000,
+    ) -> None:
+        if max_len is None:
+            max_len = max_token_count
+
+        if min_len < 0 or min_len > max_len:
+            raise ValueError("``min_len`` should be larger than 0 and equal to or smaller than ``max_len``.")
+        if max_len > max_token_count:
+            raise ValueError("``max_token_count`` must be equal to or greater than ``max_len``.")
+        datapipe = datapipe.map(partial(_token_len_fn, len_fn=len_fn))
+        if min_len is not None:
+            datapipe = datapipe.filter(partial(_token_min_filter_fn, min_len=min_len))
+        if max_len is not None:
+            datapipe = datapipe.filter(partial(_token_max_filter_fn, max_len=max_len))
```

Comment:
Super nit: Will it be more performant if we compare two steps into one filter?
```python
if min_len is not None or max_len is not None:
    datapipe.filter(partial(_token_len_filter_fn, min_len=min_len, max_len=max_len))
```

Reply:
I guess it depends on how `_token_len_filter` is implemented but it probably isn't faster.

Reply:
Updated.